azure-keyvault/azure/keyvault/custom/key_vault_authentication.py

```diff
@@ -19,27 +16,92 @@ class KeyVaultAuthBase(AuthBase):
+                # if challenge cached, use the authorization_callback to retrieve token and update the request
                 self.set_authorization_header(request, challenge)
             else:
-                # send the request to retrieve the challenge, then request the token and update
-                # the request
-                # TODO: wire up commons flag for things like Fiddler, logging, etc.
-                response = requests.Session().send(request)
-                if response.status_code == 401:
-                    auth_header = response.headers['WWW-Authenticate']
-                    if HttpBearerChallenge.is_bearer_challenge(auth_header):
-                        challenge = HttpBearerChallenge(response.request.url, auth_header)
-                        ChallengeCache.set_challenge_for_url(response.request.url, challenge)
-                        self.set_authorization_header(request, challenge)
+                # if the challenge is not cached we will let the request proceed without the auth header so we
+                # get back the proper challenge in response. We register a callback to handle the response 401 response.
+                try:
+                    self._thread_local.pos = request.body.tell()
+                except AttributeError:
+                    self._thread_local.pos = None
+
+                self._thread_local.auth_attempted = False
+                request.register_hook('response', self.handle_401)
+                request.register_hook('response', self.handle_redirect)
+
         return request
 
+    def handle_redirect(self, r, **kwargs):
+        """Reset auth_attempted on redirects."""
+        if r.is_redirect:
+            self._thread_local.auth_attempted = False
+
+    def handle_401(self, response, **kwargs):
+        """
+        Takes the response authenticates and resends if neccissary
+        
+        :return: The final response to the authenticated request
+        :rtype: requests.Response
+        """
+        # If response is not 4xx do not auth and return response
+        # See https://github.com/kennethreitz/requests/issues/3772
+        if not 400 <= response.status_code < 500:
+            self._thread_local.auth_attempted = False
+            return response
+
+        auth_header = response.headers.get('www-authenticate', '')
+
+        # if the response auth header is not a bearer challenge do not auth and return response
+        if not HttpBearerChallenge.is_bearer_challenge(auth_header):
+            self._thread_local.auth_attempted = False
+            return response
+
+        # If we've already attempted to auth for this request once, do not auth and return response
+        if self._thread_local.auth_attempted:
+            self._thread_local.auth_attempted = False
+            return response
+
+        # Otherwise authenticate and retry the request
+        self._thread_local.auth_attempted = True
+
```

Comment:
Would it make sense to move this toggle (that authn was attempted) after the successful submission of the response?